### PR TITLE
planner/cascades: add Transformation rule EliminateTableDualBelowUnionAll

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

1. This PR adds a new Transformation rule `EliminateTableDualBelowUnionAll` which will eliminates `LogicalTableDual` below `UnionAll`. This rule might be useful for partition pruning in cascades planner. 
2. Also, this PR makes some little change in the rule `EliminateProjection`.

### What is changed and how it works?

The new rule will transform:
1. `UnionAll->(X, TableDual, Y)` to `UnionAll->(X, Y)`.
2. `UnionAll->(X, TableDual)` to `X`.
3. `UnionAll->(TableDual, TableDual)` to `TableDual`.

### Check List 

Tests 

 - Unit test

Side effects

 - Possible performance regression

### CODE REVIEW

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -963,33 +966,32 @@ type EliminateProjection struct {
...
+	for i, col := range child.Prop.Schema.Columns {
 		if !col.Equal(nil, oldCols[i]) {
 			return nil, false, false, nil
 		}
 	}
 
 	// Promote the children group's expression.
-	finalGroupExprs := make([]*memo.GroupExpr, 0, child.Group.Equivalents.Len())
-	for elem := child.Group.Equivalents.Front(); elem != nil; elem = elem.Next() {
-		finalGroupExprs = append(finalGroupExprs, elem.Value.(*memo.GroupExpr))
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go]: It is not safe to reuse a `GroupExpr`. So we need to create a new `GroupExpr` with the same `LogicalPlan`.

File: planner/cascades/transformation_rules.go [hunk 2]
```diff
@@ -1952,3 +1954,72 @@ func (r *TransformAggregateCaseToSelection) isTwoOrThreeArgCase(expr expression.
...
+	newChildren := make([]*memo.Group, 0, len(expr.Children))
+	for _, childGroup := range expr.Children {
+		if !r.groupContainsTableDual(childGroup) {
+			newChildren = append(newChildren, childGroup)
+		}
+	}
+	if len(newChildren) == len(expr.Children) {
+		// There is no TableDual below UnionAll.
+		return nil, false, false, nil
+	}
```
Comment (lzmhhh123) [inline planner/cascades/transformation_rules.go]: Can we move these to function `Match`?

Comment (francis0407) [reply planner/cascades/transformation_rules.go]: The `newChildren` will be used again. If we move this to `Match`, we still have to collect the `newChildren` here.

Actually, the method `Match` is a redundant design now. We don't even need this method in the interface of `Transformation`.

**Comment by SeaRise** [comment]
In `EliminateProjection` should copy GroupExpr Instead of new GroupExpr too...

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. feel free to reopen if you want to continue with it. thank you for your contribution.

**Pull request closed without merging**